Improved error reporting

fcremo opened the pull request:
**Note**: this PR will change all the hashes, hence bumping orchestra version to 3.2.0. It also requires updating the configuration .gitignore file to ignore `.orchestra/cache` (see https://github.com/revng/orchestra/pull/38)

Relevant issues:
- https://github.com/revng/revng-orchestra/issues/24
- https://github.com/revng/revng-orchestra/issues/28

## Missing binary archives debugging

The material used to compute the component hash is now a diffable YAML, which gets saved alongside the binary archives when they are built. 
When a binary archive is not found at install time, orchestra will automatically print the diff between the current hash material and the hash material used to generate the available binary archives. There is also a plumbing command (`orc inspect component hash-material`) which prints the hash material, in case the user would want to inspect it. See the related commit message for more info.

## Other changes related to #24

- no more stack traces (for expected errors)
- failing commands are printed properly
- no more repetitions (no double printing of a failed user script, or of the exit code)

## Changes related to #28

Now `orc update` warns about components for which it could not get info from any remote. For external users this is expected (as they do not have access to the sources for e.g. revng-c).
`orc install` will fail immediately when it needs a binary archive for a component for which no HEAD commit is known.

aleclearmind commented on orchestra/actions/util/__init__.py:
```diff
@@ -105,21 +105,23 @@ def get_subprocess_output(
     :param environment: environment variables
     :param decode_as: decode the output using this encoding
     :param cwd: if not None, the command is executed in the specified path
+    :param check_returncode: by default if the subprocess returns a nonzero exit code an error is logged and an
+            InternalSubprocessException is raised. Setting check_returncode=False overrides this behaviour
```
We spent quite some time in trying to cleanup the API to spawn external process, in particular to get rid of boolean arguments.

Let's be consistent and make a `try_` method.

fcremo replied:
Now `get_script_output` (and analogues) are always used in all places where the script must either return an output and return code 0 or return an error code (in that case, an exception is raised and caught where appropriate).
The `try_*` version instead always returns the child's output, and does not check nor return the exit code.

aleclearmind replied:
Uhm, so there's no way to understand if it failed or not? Shouldn't we return a pair with the exit code?

fcremo replied:
Returning a pair is a good idea, I'll do that

aleclearmind commented on orchestra/model/component.py:
```diff
@@ -53,13 +62,52 @@ def branch(self):
+
+    @property
+    def _name_for_hash_material_cache(self) -> str:
+        return self.name.replace("/", "_").replace("-", "_")
```
Unless required for Python reasons, let's use `-` as a separator.

aleclearmind commented on orchestra/model/component.py:
```diff
@@ -53,13 +62,52 @@ def branch(self):
+
+    @property
+    def _cache_filepath(self) -> Path:
+        return Path(self._orchestra_cache_dir) / "hash_material" / self._name_for_hash_material_cache
```
So we have a file for each component? Works.
At this point, how hard is it to make it so that we can `md5sum` (or whatever hash you're using) and get as a result the name of the binary archive?

```suggestion
        return Path(self._orchestra_cache_dir) / "hash-material" / self._name_for_hash_material_cache
```

fcremo replied:
Yes, one file per component is easier and faster to work with than a single gigantic file.
Using `sha1sum` on the hash material file stored alongside the binary archive gives the recursive hash (part of the filename, which also includes the commit).

aleclearmind commented on orchestra/executor.py:
```diff
@@ -405,17 +405,9 @@ def _transitive_reduction(graph):
 
     @staticmethod
-    def _verify_binary_archives_exist(dependency_graph):
+    def _verify_prerequisites_are_met(dependency_graph):
```
```suggestion
    def _verify_prerequisites(dependency_graph):
```

aleclearmind commented:
I'm getting this error in case of missing binary archive:
```
$ orc install revng |& cleanup
Traceback (most recent call last):
  File "$ORCHESTRA_ROOT/venv/lib/python3.8/site-packages/orchestra/__init__.py", line 37, in _main
    return_code = main_parser.parse_and_execute(argv)
  File "$ORCHESTRA_ROOT/venv/lib/python3.8/site-packages/orchestra/cmds/__init__.py", line 58, in parse_and_execute
    return cmd_parser.handler(parsed_args)
  File "$ORCHESTRA_ROOT/venv/lib/python3.8/site-packages/orchestra/cmds/install.py", line 49, in handle_install
    failed = executor.run()
  File "$ORCHESTRA_ROOT/venv/lib/python3.8/site-packages/orchestra/executor.py", line 49, in run
    self._verify_prerequisites(dependency_graph)
  File "$ORCHESTRA_ROOT/venv/lib/python3.8/site-packages/orchestra/executor.py", line 410, in _verify_prerequisites
    action.assert_prerequisites_are_met()
  File "$ORCHESTRA_ROOT/venv/lib/python3.8/site-packages/orchestra/actions/install.py", line 588, in assert_prerequisites_are_met
    raise BinaryArchiveNotFoundException(self)
orchestra.exceptions.BinaryArchiveNotFoundException: Binary archive linux-x86-64/qemu/optimized/4d7d1654f5e695b7515b164b26768020fe44386b_6e4485629464f289359aae89917770da740a9c34.tar.xz for qemu@optimized not found. Try `orc update` or run `orc install -b`

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "$ORCHESTRA_ROOT/venv/bin/orc", line 8, in <module>
    sys.exit(main())
  File "$ORCHESTRA_ROOT/venv/lib/python3.8/site-packages/orchestra/__init__.py", line 49, in main
    return _main(sys.argv[1:])
  File "$ORCHESTRA_ROOT/venv/lib/python3.8/site-packages/orchestra/__init__.py", line 41, in _main
    e.log_error()
  File "$ORCHESTRA_ROOT/venv/lib/python3.8/site-packages/orchestra/exceptions.py", line 242, in log_error
    binary_archive_with_smaller_diff = min(
ValueError: min() arg is an empty sequence
```

aleclearmind commented:
I think we need to add `pygraphviz` to `requirements.txt`.

fcremo commented:
I think you were running an outdated version, I already fixed that issue (https://github.com/revng/revng-orchestra/pull/30/commits/83190cb9865521e56ccbd1ebe2c0ce512fb99f2e#diff-68613968d247c897f087450ee2ae4293ebb37195da1eaa6be734cbf2eb8cb421R241).
Re: graphviz, I added it to the dependencies, pinned at 1.6.0 because the latest 1.7.0 does not work

fcremo commented:
I did not squash https://github.com/revng/revng-orchestra/pull/30/commits/9958179e370e239fa15ac2aed411759f6d564ff9 because I would like a second opinion on it -- serializing to and parsing from raw text is not great, but it saves some appreciable time over using JSON.

aleclearmind commented:
> I did not squash 9958179 because I would like a second opinion on it -- serializing to and parsing from raw text is not great, but it saves some appreciable time over using JSON.

What do you mean? Please explain, don't just throw code at me :)

fcremo commented:
Right, here are more details.
Premise: we need to cache the material which is hashed for a given component because computing the hash material is slow (datapoint: without caching `orc components` on revng orchestra configuration takes 5s+, while taking ~0.5s with the cache). This last commit solves an issue with the caching which could result in using outdated hash material, because the commit for dependencies was not included in the cache key.
The on-disk cache stores the cache key in the first line. At first I tried JSON encoding a dictionary of { component_name: commit }, but JSON deserialization was slow enough (~1.3s invoking `orc components` like above) that I tested if manual serialization/deserialization could bring some improvement. There is an improvement, as `orc components` takes ~1.1s, but it is still taking twice as long as it did before.
I was asking opinions re: manual serialization vs JSON.

For context, the cache key now contains:
- version (so the cache can be invalidated easily if the cache key format changes)
- config hash: hash of the unprocessed yaml configuration
- a space separated list of `component_name:commit`

aleclearmind commented:
Merged in 77d576b! Thanks! Great!
